lib/em-http/http_connection.rb

```diff
@@ -40,36 +40,103 @@ def unbind(reason=nil)
+      # each certificate provided to ssl_verify_peer, fails.
+      #
+      # So, instead we remove the existing ad-hoc mechanism for verification
+      # (which did things I’m not sure it should have done, like putting
+      # non-trust-anchors into an OpenSSL::X509::Store) and instead employ
+      # OpenSSL (configured to use the system trust store, and hence able to
+      # construct the correct certification path) to do all the hard work of
+      # constructing the certification path and then verifying the peer
+      # certificate. (This is what, in my opinion, EventMachine ideally would
+      # be allowing OpenSSL to do in the first place. Instead, as far as I can
+      # tell, it pushes all of this responsibility onto its users, and then
+      # provides them with an insufficient API for meeting this
+      # responsibility.)
       def ssl_verify_peer(cert_string)
-        cert = nil
+        # We use ssl_verify_peer simply as a mechanism for gathering the
+        # certificate chain presented by the peer. In ssl_handshake_completed,
+        # we’ll make use of this information in order to verify the peer.
+        @peer_certificate_chain ||= []
         begin
           cert = OpenSSL::X509::Certificate.new(cert_string)
+          @peer_certificate_chain << cert
+          true
         rescue OpenSSL::X509::CertificateError
           return false
         end
+      end
 
-        @last_seen_cert = cert
+      def ssl_handshake_completed
+        # It’s not great to have to perform the server certificate verification
+        # after the handshake has completed, because it means:
+        #
+        # - We have to be sure that we don’t send any data over the TLS
+        #   connection until we’ve verified the certificate. Created
+        #   https://github.com/ably/ably-ruby/issues/400 to understand whether
+        #   there’s anything we need to change to be sure of this.
+        #
+        # - If verification does fail, we have no way of failing the handshake
+        #   with a bad_certificate error.
+        #
+        # Unfortunately there doesn’t seem to be a better alternative within
+        # the TLS-related APIs provided to us by EventMachine. (Admittedly I am
+        # not familiar with EventMachine.)
+        #
+        # (Performing the verification post-handshake is not new to the Ably
+        # implementation of certificate verification; the previous
+        # implementation performed hostname verification after the handshake
+        # was complete.)
+
+        # This is a certificate for which we know the server holds the private
+        # key (since the TLS connection is already established).
+        server_certificate = OpenSSL::X509::Certificate.new(get_peer_cert)
+
+        # A sense check to confirm my understanding of what’s in @peer_certificate_chain.
+        unless server_certificate == @peer_certificate_chain.last
```

Comment:
Is this necessary to have this check here? 
Because `create_certificate_store.verify(server_certificate, @peer_certificate_chain[0...-1]` checks for all certificates inside `peer_certificate_chain`. 
I mean if it's just for a normal understanding purpose, can it be moved to tests?

Reply:
Not sure I understand your question, but I’ll explain why this check is here.

I was quite worried by the description in https://github.com/eventmachine/eventmachine/issues/954#issue-1014842247 of how EventMachine "ignores all errors from the chain construction" and hence I don’t know if there is some weird scenario where, somehow, the calls to `ssl_verify_peer` terminate with some intermediate certificate instead of with the certificate of the server we’re communicating with.

That's why, instead of the previous code which passed `certificate_store.verify` the final certificate received by `ssl_verify_peer`, I explicitly use the result of `get_peer_cert`, to be sure that the certificate that we’re verifying is the one has demonstrated that they hold the private key for.

As I said, unless something has gone very wrong, these two certificates should however be identical, and hence this sense check. I don't think it does any harm being in the code, and there isn’t a place to put it "in the tests" (there aren't even any tests for SSL verification in the library, nor would it be clear where this check would sit if we did have tests).

Reply:
Would it be useful for me to add some of this explanation as a comment in the code?

Reply:
Yeah, it will be great if you can add explanation for the same !

Reply:
Also, `create_certificate_store.verify(server_certificate, @peer_certificate_chain[0...-1]` checks server certificate against all peer certificates right? 
Why don't we just check for `@peer_certificate_chain.last` if we already have check for the same at the top

Reply:
> Also, `create_certificate_store.verify(server_certificate, @peer_certificate_chain[0...-1]` checks server certificate against all peer certificates right?

It does what the code comment says:

> Verify the server’s certificate against the default trust anchors, aided by the intermediate certificates provided by the server.

(The "intermediate certificates provided by the server" are `@peer_certificate_chain[0...-1]`).

> Why don't we just check for `@peer_certificate_chain.last` if we already have check for the same right above

Do you mean "why don’t we instead write `create_certificate_store.verify(@peer_certificate_chain.last, @peer_certificate_chain[0...-1])`? If so, we could, it doesn't make a difference 🤷 Or have I misunderstood the question?

Reply:
I meant like this
```
unless create_certificate_store.verify(server_certificate, @peer_certificate_chain.last)
```

Reply:
I can be completely wrong here lol. This is based on code I understood

Reply:
See the [documentation for #verify](https://docs.ruby-lang.org/en/master/OpenSSL/X509/Store.html#method-i-verify).

- first argument is the certificate you want to verify
- second argument is an array of intermediate certificates for it to consider when constructing the certification path

i.e. the second argument needs to be all of the intermediate certificates provided by the server

Reply:
According to code above, it will ultimately match with `@peer_certificate_chain.last` right

Reply:
I don't understand the question, could you elaborate please?

Reply:
Verification always taken place against `server_certificate` and `peer_certificate_chain` right? Verification needs to be successful with just one from `peer_certificate_chain` right?

Reply:
The aim of verification is to find a chain of certificates from `server_certificate` to a certificate in the system’s trust anchors. `peer_certificate_chain` provides the verification process with a list of certificates that might aid it in constructing this chain. Does that make sense?

Reply:
> I assume you understand the implementation well 👍

I mean, I’ve done my best based on my limited knowledge, but it would be great if somebody with better TLS knowledge could take a look. I’ll see if anyone on the team can help.

Reply:
I’ll unresolve this conversation for now just because it might provide some useful info for other reviewers.

Reply:
Okay, I will resolve the conversation then !
Thanks for the review @graham-russell